packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py

```diff
@@ -0,0 +1,185 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from polywrap_core import Uri, IUriResolver, Env
+from typing import Dict, Any, List, Optional, Union
+
+@dataclass(slots=True, kw_only=True) 
+class ClientConfig():
+    """
+    This Abstract class is used to configure the polywrap client before it executes a call
+    The ClientConfig class is created and modified with the ClientConfigBuilder module
+    """
+    envs: Dict[Uri, Dict[str, Any]]
+    interfaces: Dict[Uri, List[Uri]]
+    resolver: List[IUriResolver]
+    wrappers: List[Uri]
+
+
+class BaseClientConfigBuilder(ABC):
+    """
+    An abstract base class of the `ClientConfigBuilder`, which uses the ABC module
+    to define the methods that can be used to configure the `ClientConfig` object.
+    """
+
+    def __init__(self):
+        self.config = ClientConfig(envs={}, interfaces={}, resolver=[], wrappers= [])
+
+    @abstractmethod
+    def build(self) -> ClientConfig:
+        """Returns a sanitized config object from the builder's config."""
+        pass
+
+    @abstractmethod
+    def get_envs(self)-> Dict[Uri, Dict[str, Any]]:
+        """Returns the envs dictionary from the builder's config."""
+        return self.config.envs
+
+    @abstractmethod    
+    def set_env(self, env: Env, uri: Uri):
+        """Sets the envs dictionary in the builder's config, overiding any existing values."""
+        if (env or uri) is None:
+            raise KeyError("Must provide both an env or uri")
+        self.config.envs[uri] = env
+        return self
+
+    @abstractmethod
+    def add_env(self, env: Env = None , uri: Uri = None):
```

Comment:
It shouldn't be default to None if we don't want it to be Optional

Reply:
updated